File: frontend/upload_grid.ts
import * as fs from 'fs';
import * as path from 'path';
import { DbConnection } from './client/src/module_bindings';

const DATA_FILE = path.resolve(process.cwd(), '../backend/data/manhattan.json');
const STDB_HOST = 'wss://maincloud.spacetimedb.com';
const STDB_MODULE = 'urban-whatif';

function main() {
    console.log(`Loading ${DATA_FILE}...`);
    const data = JSON.parse(fs.readFileSync(DATA_FILE, 'utf-8'));
    const grid = data.voxcity_grid;
    const width = data.shape[0];
    const depth = data.shape[1];

    console.log(`Grid dimensions: ${width} x ${depth}`);

    function normalizeVoxelCode(value: number): number {
        if (value === 0) return 0;
        if (value === -3 || value === 13) return 1;
        if (value === -2 || value === 2 || value === 5 || value === 8) return 2;
        if (value === 9) return 4;
        if (value === 1 || value === 3 || value === 4 || value === 6 || value === 7) return 5;
        if (value === 11 || value === 12 || value === 14 || value === -1) return 3;
        return value > 0 ? 3 : 0;
    }

    let surface_chars = '';
    for (let row = 0; row < width; row++) {
        for (let col = 0; col < depth; col++) {
            const column = grid[row][col];
            let top_type = 0;
            for (let level = column.length - 1; level >= 0; level--) {
                if (column[level] !== 0) {
                    top_type = normalizeVoxelCode(column[level]);
                    break;
                }
            }
            surface_chars += top_type.toString();
        }
    }

    console.log(`Generated surface string of length ${surface_chars.length}`);

    console.log('Connecting to SpacetimeDB...');
    const conn = DbConnection.builder()
        .withUri(STDB_HOST)
        .withDatabaseName(STDB_MODULE)
        .onConnect((_conn, identity) => {
            console.log('Connected! Identity:', identity.toHexString());
            console.log('Calling uploadSurfaceGrid...');
            conn.reducers.uploadSurfaceGrid({ width, depth, gridData: surface_chars });
            
            setTimeout(() => {
                console.log('Upload initiated. Closing connection.');
                process.exit(0);
            }, 3000); // Wait for the message to send
        })
        .onConnectError((_ctx, err) => {
            console.error('Connection error:', err);
            process.exit(1);
        })
        .build();
}

main();

File: backend/upload_grid.py
import json
import subprocess
from pathlib import Path

# Paths
DATA_FILE = Path(__file__).resolve().parent / "data" / "manhattan.json"
SPACETIME_APP_NAME = "urban-whatif"

def main():
    if not DATA_FILE.exists():
        print(f"Error: {DATA_FILE} not found. Run generate_city.py first.")
        return

    print("Loading manhattan.json...")
    with open(DATA_FILE, "r") as f:
        data = json.load(f)

    grid = data["voxcity_grid"]
    shape = data["shape"]
    width = shape[0]
    depth = shape[1]
    
    print(f"Grid dimensions: {width} x {depth}")
    
    # Generate 1D surface string
    # For each x, z find the highest non-zero voxel type
    surface_chars = []
    for row in range(width):
        for col in range(depth):
            column = grid[row][col]
            top_type = 0
            for level in range(len(column) - 1, -1, -1):
                if column[level] != 0:
                    top_type = column[level]
                    break
            # Convert type to char (0-9)
            # 1: Building, 2: Grass, 3: Road, 4: Water, 5: Terrain
            surface_chars.append(str(top_type))
            
    grid_string = "".join(surface_chars)
    print(f"Generated surface string of length {len(grid_string)}")
    
    # Call spacetime CLI
    print("Uploading to SpacetimeDB...")
    args_json = json.dumps({
        "width": width,
        "depth": depth,
        "gridData": grid_string
    })
    
    # Escape for shell
    escaped_json = args_json.replace('"', '\\"')
    
    cmd = f'spacetime call {SPACETIME_APP_NAME} uploadSurfaceGrid "{escaped_json}"'
    
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    if result.returncode == 0:
        print("Upload successful!")
        print(result.stdout)
    else:
        print("Upload failed!")
        print(result.stderr)

if __name__ == "__main__":
    main()
